### PR TITLE
MAINT: nd boundary grid projection. Code adapted accordingly

### PR DESCRIPTION
**Pull request by IngridKJ**
## Proposed changes

nd version of the boundary grid projection. It takes use of a function now called sparse_kronecker_product found within matrix_operations.py. This was previously a method within mortar_grid.py with the name _convert_to_vector_variable.

Some more documentation is lacking

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/numerics/linalg/matrix_operations.py [hunk 1]
```diff
@@ -834,3 +834,17 @@ def rldecode(A: np.ndarray, n: np.ndarray) -> np.ndarray:
...
     B = A[np.cumsum(j)]
     return B
+
+
+def sparse_kronecker_product(matrix: sps.spmatrix, nd: int) -> sps.spmatrix:
+    """Convert the scalar projection to a vector quantity.
+
+    If the prescribed dimension is 1, the projection matrix will in effect not be
+    altered.
+
```
Comment (IngridKJ) [inline src/porepy/numerics/linalg/matrix_operations.py:844]: Might need some documentation

**Review by IvarStefansson — COMMENTED** [review]
Thanks, Ingrid. @keileg, please have a quick look. Also, we need to wait for @Yuriyzabegaev's input before merging. Not strictly part of this PR, but we might have uncovered a previously introduced bug.

File: src/porepy/models/boundary_condition.py [hunk 2]
```diff
@@ -200,13 +200,13 @@ def dirichlet(bg: pp.BoundaryGrid):
             # Transpose to get a n_face x nd array with shape compatible with
             # the projection matrix.
             is_dir = bc_type_callable(bg.parent).is_dir.T
-            is_dir = bg.projection @ is_dir
+            is_dir = bg.projection() @ is_dir
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply src/porepy/models/boundary_condition.py:203]: `is_dir` is of shape (dim, num_cells), so it should work.

File: src/porepy/numerics/linalg/matrix_operations.py [hunk 3]
```diff
@@ -834,3 +834,22 @@ def rldecode(A: np.ndarray, n: np.ndarray) -> np.ndarray:
...
+
+def sparse_kronecker_product(matrix: sps.spmatrix, nd: int) -> sps.spmatrix:
+    """Convert the scalar projection to a vector quantity.
+
+    Used to expand projection matrices from scalar versions to a form applicable for
+    projection of nd-vector quantities.
+
+    Parameters:
+        matrix: Matrix to be expanded using a kronecker product.
```
Comment (keileg) [inline src/porepy/numerics/linalg/matrix_operations.py:846]:
```suggestion
        matrix: Matrix to be expanded using a Kronecker product.
```

File: src/porepy/numerics/linalg/matrix_operations.py [hunk 4]
```diff
@@ -834,3 +834,22 @@ def rldecode(A: np.ndarray, n: np.ndarray) -> np.ndarray:
     j[i[1:-1:]] = 1
     B = A[np.cumsum(j)]
     return B
+
+
+def sparse_kronecker_product(matrix: sps.spmatrix, nd: int) -> sps.spmatrix:
```
Comment (keileg) [inline src/porepy/numerics/linalg/matrix_operations.py:839]: Do we need a test for this? I'm inclined to say yes, but testing 2-3 cases should be sufficient.

**Review by keileg — COMMENTED** [review]
Nice work, one comment only, but also a question: Have you explicitly decided we need no tests for this function?

File: src/porepy/numerics/linalg/matrix_operations.py (hunk 4 above)
Comment (IvarStefansson) [reply src/porepy/numerics/linalg/matrix_operations.py:839]: What would you test?

File: src/porepy/params/bc.py [hunk 5]
```diff
@@ -241,13 +241,13 @@ class BoundaryConditionVectorial(AbstractBoundaryCondition):
         num_faces (int): Number of faces in the grid of the subdomain.
         dim (int): Dimension of the boundary. One less than the dimension of
             the subdomain.
-        is_neu (np.ndarray boolean, size sd.dim x sd.num_faces): Element i is true if
-            face i has been assigned a Neumann condition. Tacitly assumes that
-            the face is on the boundary. Should be false for internal faces, as
-            well as Dirichlet faces.
-        is_dir (np.ndarray, boolean, size sd.dim x sd.num_faces): Element i is true if
+        is_neu (np.ndarray boolean, shape(sd.dim, sd.num_faces)): Element i is true if
```
Comment (IvarStefansson) [inline src/porepy/params/bc.py:244]:
```suggestion
        is_neu (np.ndarray boolean, shape=(sd.dim, sd.num_faces)): Element i is true if
```

File: src/porepy/params/bc.py [hunk 6]
```diff
@@ -241,13 +241,13 @@ class BoundaryConditionVectorial(AbstractBoundaryCondition):
...
+            face i has been assigned a Neumann condition. Tacitly assumes that the face
+            is on the boundary. Should be false for internal faces, as well as Dirichlet
+            faces.
+        is_dir (np.ndarray, boolean, shape=(sd.dim sd.num_faces)): Element i is true if
```
Comment (IvarStefansson) [inline src/porepy/params/bc.py:248]:
```suggestion
        is_dir (np.ndarray, boolean, shape=(sd.dim, sd.num_faces)): Element i is true if
```

**Review by IvarStefansson — APPROVED** [review]

**Pull request merged**